sleepy/types.py

```diff
@@ -656,6 +671,19 @@ def children(self) -> Collection[Type]:
   def has_same_symbol_as(self, other: Type) -> bool:
     return isinstance(other, UnionType)
 
+  def can_be_interpreted_as(self, other: Type) -> bool:
+    if not isinstance(other, UnionType):
+      return False
+    if self.val_size != other.val_size:
+      return False
+    for own_variant_type, variant_index in self.variant_index_map.items():
+      if variant_index not in other.variant_index_map.values():
+        return False
+      other_variant_type = other.get_variant_at_index(variant_index)
+      if not own_variant_type.can_be_interpreted_as(other_variant_type):
+        return False
+    return True
+
```

Comment:
Is this equivalent to the following?
```suggestion
  def can_be_interpreted_as(self, other: Type) -> bool:
    return (isinstance(other, UnionType) and
            self.val_size == other.val_size and
            # all type, index pairs of self exist in other
            all(other.variant_index_map.get(variant) == index for variant, index in self.variant_index_map.items()))
```

Reply:
No, it's not quite, because we do not require our own variant types to be equal to the other variant type with the same num, but instead that the our variant type can be interpreted as the other recursively.
e.g. `T|Ref[A|B]` shares the memory layout with `T|Ref[A|B|C]`.

Reply:
Okay, so this instead

```suggestion
  def can_be_interpreted_as(self, other: Type) -> bool:
      return (isinstance(other, UnionType) and
            self.val_size == other.val_size and
            # all type, index pairs of self have a corresponding one in other
            all(
              index in other.possible_type_nums and other.get_variant_at_index(index).can_be_interpreted_as(variant)
              for variant, index in self.variant_index_map.items()))
```